R/do_spark.R

```diff
@@ -0,0 +1,126 @@
+    unnamed <- ! nzchar(optnames)
+    if (any(unnamed)) {
+      warning('ignoring doSpark package option(s) specified with unnamed argument')
+      opts <- opts[!unnamed]
+      optnames <- optnames[!unnamed]
+    }
+
+    # filter out unrecognized options with a warning
+    recog <- optnames %in% .valid_options
+    if (any(!recog)) {
+      warning(sprintf('ignoring unrecognized doSpark package option(s): %s',
+                      paste(optnames[!recog], collapse = ', ')), call. = FALSE)
+      opts <- opts[recog]
+      optnames <- optnames[recog]
+    }
+
+    # clear .options from any previous call to registerDoSpark
+    remove(list = ls(.options), pos = .options)
+
+    # set new options
+    for (i in seq(along=opts)) {
+      assign(optnames[i], opts[[i]], pos=.options)
+    }
+  }
+
+  # internal function called by foreach
+  .doSpark <- function(obj, expr, envir, data) {
+    # internal function to compile an expression if possible
+    .compile <- function(expr, ...) {
+      if (getRversion() < "2.13.0" || isTRUE(.options$nocompile))
+        expr
+      else
+        compiler::compile(expr, ...)
+    }
+
+    # internal functions to serialize and unserialize an arbitrary R object to/from string
+    #
+    # NOTE: here we are (reasonably) assuming foreach items are not astronomical in size and
+    # therefore the ~33% space overhead from base64 encode is still acceptable.
+    # If this assumption were not true, then base64 would need to be replaced with another
+    # more efficient binary-to-text encoding such as yEnc (which has only 1-2% space overhead).
+    .encode_item <- function(item) list(encoded = base64enc::base64encode(serialize(item, NULL)))
+    .decode_item <- function(item) unserialize(base64enc::base64decode(item))
+
+    # internal function to process spark data frames
+    .process_spark_items <- function(...) {
+      f <- function(item) {
+        # NOTE: if `expr` involves any object outside of its own environment then this current
+        # implementation will *attempt* to serialize and send those object(s) to Spark workers
+        # as well -- e.g., consider the following:
+        #
+        # foo <- 5
+        # foreach (bar = 1:100) %dopar% { foo * bar }  # `foo` is outside of `expr`'s environment
+        #
+        # However, last time I checked https://spark.rstudio.com/guides/distributed-r this is
+        # not a supported use case yet
+        # (see https://spark.rstudio.com/guides/distributed-r/#closures)
+        res <- eval(expr, envir=.decode_item(item$encoded), enclos=envir)
+        .encode_item(res)
+      }
+      encoded_res <- sparklyr::sdf_collect(spark_items %>% sparklyr::spark_apply(f, ...))[[1]]
+      lapply(encoded_res, .decode_item)
+    }
+
+    if (!inherits(obj, "foreach"))
+      stop("obj must be a foreach object")
+
+    spark_conn <- data$spark_conn
+    spark_apply_args <- data$spark_apply_args
+
+    it <- iterators::iter(obj)
```

Comment:
Most likely not, because the accumulator functionality in `foreach` is implemented in term of iterator:

```
> makeAccum                                                                                                                                                                                                 function (it)                                                                                                                                                                                                      
{                                                                                                                                                                                                                  
    function(results, tags) {                                                                                                                                                                                      
        if (identical(it$error.handling, "stop") && !is.null(it$state$errorValue)) 
            return(invisible(NULL))
        for (i in seq(along.with = tags)) {
            if (it$verbose) 
                cat(sprintf("got results for task %d\n", tags[i]))
            accumulate(it, results[[i]], tags[i])
        }
    }
}
```

However, because `iterators` is definitely a dependency of the `foreach` package already, I think all we have to do is to list `foreach` as a suggested dependency for `sparklyr` so that `iterators` package will not be automatically installed when installing `sparklyr`.